codex/node.nim

```diff
@@ -415,13 +436,21 @@ proc start*(node: CodexNodeRef) {.async.} =
       # since fetching of blocks will have to be selective according
       # to a combination of parameters, such as node slot position
       # and dataset geometry
-      if fetchErr =? (await node.fetchBatched(manifest, onBatch = onBatch)).errorOption:
+      if fetchErr =? (await node.fetchBatched(manifest, onBatch = onBatch, expiry = some request.expiry.toSecondsSince1970)).errorOption:
         let error = newException(CodexError, "Unable to retrieve blocks")
         error.parent = fetchErr
         return failure(error)
```

Comment:
I find it a bit odd to be updating a block's expiry inside of `fetchBatched`. I would imagine a caller could call `fetchBatched` without understanding that it would iterate all blocks and update their expiry. Perhaps it might be better to have `fetchBatched` return `seq[Blocks]` and then `ensureExpiry` called on that sequence after `fetchBatched` completes?
```nim
without blocks =? (await node.fetchBatched(manifest, onBatch = onBatch)), fetchErr:
  let error = newException(CodexError, "Unable to retrieve blocks")
  error.parent = fetchErr
  return failure(error)

await allFuturesThrowing(blocks.mapIt(node.blockStore.ensureExpiry(it.read.get.cid, request.expiry.toSecondsSince1970)))
```

Additionally, using `.get.cid` may crash the node with a defect if `it.read` doesn't contain a block, so possibly could flesh this out a bit more, which may also help with readability:
```nim
await allFuturesThrowing(
  blocks.map(
    proc(blok: Future[?!Block]): Future[?!void] = (
      without blk =? await blok, err:
        return failure(err)
      return node.blockStore.ensureExpiry(blk.cid, request.expiry.toSecondsSince1970)
    )
  )
)
```

Reply:
> This design comes from the limitation of the BlockStore.getBlock() abstraction. The "correct" approach is calling putBlock() with the expected TTL, but because of how NetworkStore works, it actually calls putBlock during getBlock (eq. when it fetches data from network) and because of the interface we can't specify TTL for the getBlock.

But sticking it into `fetchBatched` isn't addressing this particular limitation at all it simply adds an unexpected side effect to `fetchBatched`. What @emizzle is suggesting is the correct approach here.

As for the perceived limitation of the `network` store interface, an explicit specialization with an additional `method` or `proc`, for example `getOrFetch(cid, ttl)`, that `get`s from the local store or `fetches` from the network, would address this quite nicely.

Reply:
I don't perceive it as a problem. In my mind, `fetchBatched` is for "downloading data from the network" and the expiry extends this premise to "download data from network and store it for X time" which is IMHO tightly coupled. Btw. if you don't set `expiry` nothing is set... The default expiry is `none` so by no behavior is changed from the original one.

That said, I don't mind changing it as you suggested, but my question then is, how should I address this problem?
>There is also a potential problem that if you are downloading some very big dataset that might exceed the DefaultTTL (currently 24 hours), the maintenance could start deleting the data from the beginning of the dataset before you finish the download. Hence that is why the ensureExpiry is called in the onBatch callback.

If we bring up the expected dataset size (or slot size? not sure) of 1 TB, then if the download is slower than 11MB/s  (`1 000 000MB / (24*60*60)`) then we will run into this problem, which is IMHO quite plausible. So should I move the expiry update call into the Marketplace's `onBatch` implementation? Or should I ignore this, call it edge-case and update only once `fetchBatched` is finished?

Also, keep in mind that the `DefaultTTL` is configurable by the user through CLI parameter...

Reply:
> I don't perceive it as a problem. In my mind, fetchBatched is for "downloading data from the network" and the expiry extends this premise to "download data from network and store it for X time" which is IMHO tightly coupled. Btw. if you don't set expiry nothing is set... The default expiry is none so by no behavior is changed from the original one.

The point is that there is no way to tell that there is some ttl related logic going on during download, this isn't explicit at all and it's unexpected behavior, overall this is a bad design choice. 

> There is also a potential problem that if you are downloading some very big dataset that might exceed the DefaultTTL (currently 24 hours), the maintenance could start deleting the data from the beginning of the dataset before you finish the download. Hence that is why the ensureExpiry is called in the onBatch callback.

> So should I move the expiry update call into the Marketplace's onBatch implementation? 

No need for that, just create another `onBatch` in the `onStore` callback update the expiry and the call the marketplace `onBatch` after that. 

Like so:

```nim
hostContracts.sales.onStore = proc(request: StorageRequest,
                                    slot: UInt256,
                                    onBatch: BatchProc): Future[?!void] {.async.} =

  proc myOnBatch(blocks: seq[bt.Block]) = 
    # update expiry here
    blocks.mapIt( store.ensureExpiry(...) )
   onBatch()
  

  if fetchErr =? (await node.fetchBatched(manifest, onBatch = myOnBatch)).errorOption:
    let error = newException(CodexError, "Unable to retrieve blocks")
    error.parent = fetchErr
    return failure(error)

```

> Or should I ignore this, call it edge-case and update only once fetchBatched is finished?

Definitely no, I think at the very least the above works fine.

Alternatively, you could also make the network store final and add a `getOrFetch(cid, ttl)` function - this is probably the cleanest solution and it addresses the limitations you're facing. The top level consumers of the interface probably only care about this use either way.

-----------


Also, this:

```nim
  let iter = Iter.fromSlice(0..<manifest.blocksCount)
    .map((i: int) => node.blockStore.getBlock(BlockAddress.init(manifest.treeCid, i)))

  for batchNum in 0..<batchCount:
    let blocks = collect:
      for i in 0..<batchSize:
        if not iter.finished:
          iter.next()
```

Looks awfully weird, I understand whats going on, but we probably want to add a `collect` call or something similar to the async iterators for this types of use cases.

Reply:
Great! Thanks a lot for the input. I will put together fix tomorrow 🙏

Reply:
> Alternatively, you could also make the network store final and add a getOrFetch(cid, ttl) function - this is probably the cleanest solution and it addresses the limitations you're facing. The top level consumers of the interface probably only care about this use either way.

Agree that this would be best, but IMHO lets go with the simplest now. We can refactor this later.